modules/core/src/weaver/MemoisedResource.scala

```diff
@@ -0,0 +1,65 @@
+package weaver
+
+import cats.effect._
+import cats.syntax.all._
+
+import CECompat.{ Deferred, Ref }
+
+object MemoisedResource {
+  def apply[F[_]: Concurrent, A](
+      resource: Resource[F, A]): F[Resource[F, A]] =
+    new MemoisedResource[F, A].apply(resource)
+}
+
+private class MemoisedResource[F[_]: Concurrent, A] {
+
+  sealed trait State
+  case object Uninitialised extends State
+  case class InUse(
+      value: Deferred[F, Either[Throwable, A]],
+      finalizer: F[Unit],
+      uses: Int)
+      extends State
+
+  def apply(resource: Resource[F, A]): F[Resource[F, A]] =
+    Ref[F].of[State](Uninitialised).map { ref =>
+      val initialise: F[A] = for {
+        valuePromise     <- Deferred[F, Either[Throwable, A]]
+        finaliserPromise <- Deferred[F, F[Unit]]
+        compute <- ref.modify {
+          case Uninitialised =>
+            val newState = InUse(valuePromise, finaliserPromise.get.flatten, 1)
+            val compute = Concurrent[F].attempt(resource.allocated).flatMap {
+              case Right((a, fin)) => for {
+                  _ <- valuePromise.complete(Right(a))
+                  _ <- finaliserPromise.complete(fin)
+                } yield a
+              case Left(e) => for {
+                  _ <- valuePromise.complete(Left(e))
+                  _ <- finaliserPromise.complete(Concurrent[F].unit)
+                  a <- Concurrent[F].raiseError[A](e)
+                } yield a
+            }
+            newState -> compute
+          case InUse(value, finalizer, uses) =>
+            val newState      = InUse(value, finalizer, uses + 1)
+            val compute: F[A] = value.get.flatMap(Concurrent[F].fromEither)
+            newState -> compute
+        }
+        value <- compute
+      } yield value
+
+      val finalise: F[Unit] = ref.modify[F[Unit]] {
+        case Uninitialised =>
+          Uninitialised -> Concurrent[F].raiseError(
+            new IllegalStateException("Implementation error"))
+        case InUse(_, finaliser, n) if n <= 1 =>
+          Uninitialised -> finaliser
+        case InUse(value, finaliser, n) =>
+          InUse(value, finaliser, n - 1) -> Concurrent[F].unit
+      }.flatten
+
+      Resource.make(initialise)(_ => finalise)
```

Comment:
Resource allocation and release are both uncancelable in CE2 so this is cancellation-safe

Reply:
In CE3 the `make` constructor is also uncancelable, but the underlying `Resource` that is `allocated` may be cancelable so may be losing some cancellation timeliness by doing this. might be good to explore that further (if you even need it, that is :P )

Reply:
Actually, the above is also true for CE2: you can cancel in between multiple `Resource` acquisitions, but not within a single acquisition. There's no point in trying to get around that in CE2 though :P

Reply:
That's great, thanks for having a look ! I'm gonna have a look at makeFull, it's probably worth specialising the implementations.

Because the construct is provided for the user to share resources between test suites (via indirection), I'd rather ensure that it's cancellation-safe, as I have no clue what people might want to be doing with this capability. 

 I might punt that to a later point though.